### DISCUSSION TITLE
FileStat permissions is always undefined for readonly files

### DISCUSSION THREAD

**Discussion by ArturoDent** [microsoft/vscode-discussions#673] [Extension Development QnA] [answered]
@gjsjohnmurray 

Has `vscode.FileStat` regressed with respect to permissions? If I try: 

```ts
    const stat: vscode.FileStat = await vscode.workspace.fs.stat(uri);

    if (stat.permissions === vscode.FilePermission.Readonly) {}
```

`stat.permissions` is always undefined although I set the permissions to `readonly` in the OS filesystem and the file correctly has the lock icon in vscode-insiders.

**Comment by gjsjohnmurray** [comment]
I'm not aware of any, but I guess the recent work by @jackpunt and @bpasero may have affected this.

**Comment by bpasero** [comment, marked as answer]
Please first figure out if this used to work and now does not work anymore. As far as I can tell, we never reported `FilePermission.Readonly` for files that are readonly on disk.

To help us narrow down what change caused the issue, please run the following command to find the build. It will start previous releases of VS Code insiders to see where the regression was introduced:
* `npx --yes vscode-bisect@latest`
* follow the instructions until you found the offending build
* report back the commit range in this issue

**Reply by jackpunt** [reply]
I agree: Previously, DiskFileProvider explicitly/intentionally did not set Readonly from the file stat.
(which was the impetus for this whole chain of mods to more correctly report file stat, 
  while still retaining an ability to edit & over-write when desired)

see comment https://github.com/microsoft/vscode/issues/17670#issuecomment-1405367866

**Comment by ArturoDent** [comment]
@jackpunt 

I assume you are saying that for a file set to read-only in the OS, that `FileStat.permissions` is not set to anything, it remains `undefined`.  That is what I see.

When I set a file to `read-only` in vscode, via the `File: Set Active Editor Read-only in Session` command and then check that file's permissions via 

```ts
	const activeUri = vscode.window.activeTextEditor.document.uri;

	const stats = await vscode.workspace.fs.stat(activeUri);
```

The `stats.permissions` object is also `undefined` - is that expected?  When is `FilePermission.Readonly` set?

**Reply by bpasero** [reply]
Here is the condition that makes `FilePermission.Readonly` to be set:
* the entire file system provider is marked as "readonly"
* the file is marked as "readonly" by the file system provider

Here are conditions that do not impact `FilePermissions`:
* the file is marked as "readonly" by the operating system (e.g. via file permissions)
* the file is marked as "readonly" via configuration by the user (essentially the new feature we introduced)

As such, today extensions cannot figure out any of the above 2 conditions yet. Please file a feature request for it.
